NationStatesAPIBot/NationStatesAPIBot/Services/DumpDataService.cs

```diff
@@ -10,47 +10,181 @@
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 using NationStatesAPIBot.Types;
+using System.IO;
 
 namespace NationStatesAPIBot.Services
 {
     public class DumpDataService
     {
         private readonly ILogger<DumpDataService> _logger;
-        private readonly BaseApiService _apiService;
-        private static List<NATION> _nations;
-        private static List<REGION> _regions;
-        private static Task _updating;
-
-        public DumpDataService(ILogger<DumpDataService> logger, BaseApiService apiService, CancellationToken stop)
+        private readonly NationStatesApiService _apiService;
+        private List<NATION> _nations;
+        private List<REGION> _regions;
+        private bool isDumpUpdateCycleRunning = false;
+        private CancellationTokenSource tokenSource = new CancellationTokenSource();
+        private DateTime lastDumpUpdateTime = DateTime.UnixEpoch;
+        private string regionFileName = "regions-dump-latest.xml.gz";
+        private string nationFileName = "nations-dump-latest.xml.gz";
+        private EventId defaultEventId;
+        public DumpDataService(ILogger<DumpDataService> logger, NationStatesApiService apiService, CancellationToken stop)
         {
             _logger = logger;
             _apiService = apiService;
-            _logger.LogInformation($"--- DumpDataService started ---");
-            PeriodicUpdate(TimeSpan.FromDays(1), stop);
+            defaultEventId = LogEventIdProvider.GetEventIdByType(LoggingEvent.DumpDataServiceAction);
+            _logger.LogInformation(defaultEventId, GetLogMessage("--- DumpDataService started ---"));
+            StartDumpUpdateCycle();
+        }
+
+        public bool IsUpdating { get; private set; } = false;
+
+        public bool DataAvailable { get; private set; } = false;
+
+        private string GetLogMessage(string message)
+        {
+            return LogMessageBuilder.Build(defaultEventId, message);
+        }
+
+        public void StartDumpUpdateCycle()
+        {
+            isDumpUpdateCycleRunning = true;
+            Task.Run(async () => await UpdateCycle());
         }
-        
-        public async Task PeriodicUpdate(TimeSpan interval, CancellationToken stop)
+
+        public void StopDumpUpdateCycle()
         {
-            while (true)
+            isDumpUpdateCycleRunning = false;
+            tokenSource.Cancel();
+        }
+
+        private async Task UpdateCycle()
+        {
+            _logger.LogInformation(defaultEventId, GetLogMessage("--- DumpDataService Update Cycle has been started ---"));
+            while (isDumpUpdateCycleRunning)
             {
-                await Update();
-                await Task.Delay(interval, stop);
+                await UpdateData();
```

Comment:
Does this loop need a delay, or is there a delay somewhere along the line?

Reply:
In UpdateData should be a WaitForAction. But we can add a small delay too though.

Reply:
Ah yes, now I see how that works.